### ISSUE TITLE
transient or private functions in Endpoints

### ISSUE THREAD

**Issue by starklord** [serverpod/serverpod#1245]
**Is your feature request related to a problem? Please describe.**
What is the way to create transient or API functions inside Endpoints?

**Describe the solution you'd like**
Allow the creation of private functions inside endpoints that will not be public to the client.

**Additional context**

**Comment by SandPod** [comment]
Hey 👋 
If I understand your question correctly you would like to define a server endpoint but not have the client side library generated.

Right now we don't support this feature but it could be a good addition for the future.

**Comment by realmeylisdev** [comment]
This has shipped since the issue was filed — current Serverpod supports exactly what's requested here, in two ways. Everything below is verified against `main` at [`43c4e0e`](https://github.com/serverpod/serverpod/commit/43c4e0eb14917882447fe47b9efe9dc199c329e5).

### 1. Private Dart methods are never exposed

The endpoint analyzer skips private methods — and any method that doesn't take a `Session` as its first parameter — so a `_helper` inside an endpoint produces no client bindings and no server dispatch:

```dart
static bool isEndpointMethod(MethodElement method) {
  if (method.isPrivate) return false;
  if (method.markedAsIgnored) return false;

  if (_excludedMethodNameSet.contains(method.name)) return false;

  return method.formalParameters.isFirstRequiredParameterSession;
}
```

– [`endpoint_method_analyzer.dart#L79-L88`](https://github.com/serverpod/serverpod/blob/43c4e0eb14917882447fe47b9efe9dc199c329e5/tools/serverpod_cli/lib/src/analyzer/dart/endpoint_analyzers/endpoint_method_analyzer.dart#L79-L88)

This is pinned by a dedicated test asserting that a private method produces no method definition at all: [`endpoint_method_test.dart#L914-L950`](https://github.com/serverpod/serverpod/blob/43c4e0eb14917882447fe47b9efe9dc199c329e5/tools/serverpod_cli/test/integration/analyzer/dart/endpoint_validation/endpoint_method_test.dart#L914-L950)

### 2. `@doNotGenerate` for public methods (or whole endpoint classes)

If the helper should stay public on the Dart level (e.g. shared with other endpoints or server code) but hidden from the protocol:

```dart
class ExampleEndpoint extends Endpoint {
  Future<String> hello(Session session, String name) async => 'Hello $name';

  @doNotGenerate
  Future<String> helper(Session session) async => 'not exposed';
}
```

No client code and no server dispatch are generated for the annotated method (or class); it remains an ordinary Dart method.

- Annotation definition: [`packages/serverpod_shared/lib/src/annotations.dart#L3-L9`](https://github.com/serverpod/serverpod/blob/43c4e0eb14917882447fe47b9efe9dc199c329e5/packages/serverpod_shared/lib/src/annotations.dart#L3-L9)
- Docs: [Ignore endpoint definition](https://docs.serverpod.dev/concepts/working-with-endpoints#ignore-endpoint-definition)
- History: introduced as `@ignoreEndpoint` for whole endpoint classes in #2954 (first shipped in 2.3.0, December 2024), extended to individual methods in #3404 (2.6.0, April 2025 — [CHANGELOG](https://github.com/serverpod/serverpod/blob/43c4e0eb14917882447fe47b9efe9dc199c329e5/CHANGELOG.md#L553)), then renamed to `@doNotGenerate`, deprecating the old name, in #3484 (2.7.0, May 2025 — [CHANGELOG](https://github.com/serverpod/serverpod/blob/43c4e0eb14917882447fe47b9efe9dc199c329e5/CHANGELOG.md#L544))
- Generator behavior pinned by [`endpoint_do_not_generate_annotation_test.dart`](https://github.com/serverpod/serverpod/blob/43c4e0eb14917882447fe47b9efe9dc199c329e5/tools/serverpod_cli/test/integration/analyzer/dart/endpoint_validation/endpoint_do_not_generate_annotation_test.dart) — an annotated endpoint produces no endpoint definition, and when two endpoints share a file only the non-annotated one is generated

You can also see it working in this repo's own test fixtures: [`AbstractBaseEndpoint.ignoredMethod`](https://github.com/serverpod/serverpod/blob/43c4e0eb14917882447fe47b9efe9dc199c329e5/tests/serverpod_test_server/lib/src/endpoints/endpoint_inheritance.dart#L10-L14) is annotated with `@doNotGenerate`, and the [generated client class for that endpoint](https://github.com/serverpod/serverpod/blob/43c4e0eb14917882447fe47b9efe9dc199c329e5/tests/serverpod_test_client/lib/src/protocol/client.dart#L989-L1002) contains the un-annotated methods (`virtualMethod`, `abstractBaseMethod`, …) but no `ignoredMethod`; likewise the server dispatcher's `concreteBase` connector block has no `ignoredMethod` entry. (If you grep the generated files you will find one `ignoredMethod` — under `moduleEndpointExtension`, a fixture that deliberately re-exposes it via an un-annotated override; see the per-declaration semantics below.)

Two details worth knowing:

- The annotation applies per declaration: a subclass can deliberately re-expose an inherited hidden method by overriding it without the annotation ([docs](https://docs.serverpod.dev/concepts/working-with-endpoints#unhiding-endpoint-methods-annotated-with-donotgenerate-in-the-super-class)).
- `@doNotGenerate` removes the server route as well as the client code. Keeping the server route while hiding only the client binding is a separate open request: #3479.

So this looks like it can be closed as completed (method-level hiding available since 2.6.0, under the current `@doNotGenerate` name since 2.7.0).

**Comment by marcelomendoncasoares** [comment]
Indeed, the `@doNotGenerate` is the fix for this issue! Thanks again for helping us clean up the backlog, @realmeylisdev 🎉
